### PR TITLE
Add initial support for a cleaner interface for the first order dynamics model 

### PR DESCRIPTION
**Pull request by apozharski**
This adds support for the object model described in the RSS FOR short paper.

### CODE REVIEW

File: +nosnoc/+dcs/Base.m [hunk 1]
```diff
@@ -37,19 +37,19 @@
             names = properties(obj);
             for ii=1:length(names)
                 name = names{ii};
-                if ~endsWith(name, '_fun')
+                if ~endsWith(name, '_fun') || isa(obj.(name), 'double')
```
Comment (apozharski) [inline +nosnoc/+dcs/Base.m:40]: This is a bugfix which caused errors when trying to print dcs

File: +nosnoc/+dcs/Objects.m [hunk 2]
```diff
@@ -0,0 +1,128 @@
+classdef Objects < nosnoc.dcs.Base
+% A particular class of dynamic complementarity system called a gradient complementarity system.
```
Comment (nurkanovic) [inline +nosnoc/+dcs/Objects.m:2]: this is the description of the GCS class. probably a copy past leftover, please provide better description

File: +nosnoc/+dcs/Objects.m [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    properties
+        lambda % casadi.SX|casadi.MX: Lagrange multipliers for the gradient complementarity system.
+        c_lift % casadi.SX|casadi.MX: Lift variables for the gap functions in :class:`nosnoc.model.Pds`.
+
+        x_dot_lift
+        g_friction
+        
+        f_x % casadi.SX or casadi.MX: Right hand side of the gradient complementarity system
+
+        dims % struct: dimensions struct TODO document members
```
Comment (nurkanovic) [inline +nosnoc/+dcs/Objects.m:21]: has everything as the GCS and more, and objects is not a name of a particular DCS.
please explain what is the motivation for this, and better document the class by explaining briefly what is it doing, and what it the difference to GCS

File: +nosnoc/+model/Objects.m [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+        % x_dot lift
+        x_dot_lift
+
+        % friction expressions
+        G_friction
+        H_friction
+        g_friction
+
+        % Feasible set
+        r = 0.01
```
Comment (nurkanovic) [inline +nosnoc/+model/Objects.m:51]: ?

File: +nosnoc/+model/Objects.m [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+        normal
+        normal0
+        tangent
+        tangent0
+
+        % Distance functions
+        c
+
+        % objects
+        objects
```
Comment (nurkanovic) [inline +nosnoc/+model/Objects.m:63]: the boring part of work, but would be of great help if every property has a one line description.
again the name objects is a bit out of line. what about having everything a gcs
or call it gcs_objects, pds_objects, gcs_sdf_objects .. something that is more consistent

File: +nosnoc/+model/Objects.m [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+
+        function verify_and_backfill(obj, opts)
+            import casadi.*
+            verify_and_backfill@nosnoc.model.Base(obj,opts);
+
+            dims = obj.dims;
+            
+            dims.n_c = size(obj.c, 1);
+            dims.n_lambda = dims.n_c;
+            dims.n_y1d = size(obj.y1_d, 1);
```
Comment (nurkanovic) [inline +nosnoc/+model/Objects.m:81]: ?
described what is what, otherwise no one knows how to interpret the variables and their features

File: +nosnoc/+model/Objects.m [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+            
+            % update model
+            obj.c = vertcat(obj.c,c);
+            obj.lambda = vertcat(obj.lambda, lambda);
+        end
+
+        function addEllipseBall(obj, ellipse, ball, mu_f)
+            import casadi.*
+            n_dim = ball.n_dim;
+
```
Comment (nurkanovic) [inline +nosnoc/+model/Objects.m:206]: every function would benefit greatly of a 1-2 line descripiton (most names are self explaining, slightly more details would be welcome

File: +nosnoc/+model/Objects.m [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+                R = Function('R', {theta}, {R_matrix});
+                a = SX.sym('a',2);
+                b = SX.sym('b',2);
+                cross_fun = Function('cross', {a,b}, {a(1)*b(2) - a(2)*b(1)});
+            else
+                rx = SX.sym('rx',1);
+                ry = SX.sym('ry',1);
+                rz = SX.sym('rz',1);
+
+                Rot = [cos(ry)*cos(rz), sin(rx)*sin(ry)*cos(rz) - cos(rx)*sin(rz),  cos(rx)*sin(ry)*cos(rz) + sin(rx)*sin(rz);
```
Comment (nurkanovic) [inline +nosnoc/+model/Objects.m:221]: describe

File: +nosnoc/+model/Objects.m [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+            if n_dim == 2
+                theta = SX.sym('theta');
+                R_matrix = [cos(theta) -sin(theta);...
+                    sin(theta) cos(theta)];
+                R = Function('R', {theta}, {R_matrix});
+                a = SX.sym('a',2);
+                b = SX.sym('b',2);
+                cross_fun = Function('cross', {a,b}, {a(1)*b(2) - a(2)*b(1)});
+            else
+                rx = SX.sym('rx',1);
```
Comment (nurkanovic) [inline +nosnoc/+model/Objects.m:217]: different naming convtion for 2 and 3d?

File: +nosnoc/+model/Objects.m [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+
+                L = alpha - 1 + mu'*g_d;
+                
+                % g_kkt = [1 - mu_1c - mu_2c;
+                %     mu_1c*(R(ellipse.xi)*(A + A')*R(ellipse.xi)')*(p_d-ellipse.c) + mu_2c*(2/r^2)*(p_d-ball.c)];
+                g_kkt = L.jacobian([alpha;p_d])';
+                obj.g_kkt = vertcat(obj.g_kkt, g_kkt);
+
+                
+                ntr = 2/ball.r^2*(p_d - ball.c)*mu_2c;
```
Comment (nurkanovic) [inline +nosnoc/+model/Objects.m:259]: comments, comments, comments

File: +nosnoc/+model/Objects.m [hunk 11]
```diff
@@ -0,0 +1,421 @@
...
+                normal_ball = [-ntr];
+                normal_ellipse =[ntr;
+                    cross_fun(p_d-ellipse.c, ntr)];
+
+                obj.normal = [obj.normal;normal_ball;normal_ellipse];
+                obj.normal0 = [obj.normal0;0;1;0;-1;0];
```
Comment (nurkanovic) [inline +nosnoc/+model/Objects.m:265]: ?

File: object_app/objects_main.m [hunk 12]
```diff
@@ -0,0 +1,350 @@
+function objects_main
```
Comment (nurkanovic) [inline object_app/objects_main.m:1]: is there an example of use?

File: examples/object_model/ellipses.m [hunk 13]
```diff
@@ -0,0 +1,90 @@
+clear all
```
Comment (nurkanovic) [inline examples/object_model/ellipses.m:1]: as this example introduces this new feature, commeting every line of the new options would be helpful.
currently requires lot of trail and error to figure out what is what  - please make the user's life easy.

also ma27 has a lot of trouble with this example on my machine in the first homotopy iteration. dual inf goes to 1e14. later iterations are easy. 
is the problem well scaled?

**Comment by nurkanovic** [comment]
The code looks elegant; however, the complete absence of comments both on the math equation parts and later in the examples makes it very hard to follow what is done in particular. 

before it can be merged, needs lot of one line comments to make it more sustainable and readable. currently only you understand what is what.

File: examples/object_model/ellipses.m (hunk 13 above)
Comment (apozharski) [reply examples/object_model/ellipses.m:1]: Interesting... ma27 converges pretty well on my machine (though in the initial phase dual infeasibility does go to 1e8 but that is not uncommon in my experience). Will add comments.

**Comment by apozharski** [comment]
# Summary - nosnoc coverage
<details open><summary>Summary</summary>

|||
|:---|:---|
| Generated on: | 11/22/2024 - 10:10:43 |
| Coverage date: | 11/22/2024 - 10:10:34 |
| Parser: | Cobertura |
| Assemblies: | 12 |
| Classes: | 51 |
| Files: | 51 |
| **Line coverage:** | 41.5% (2830 of 6803) |
| Covered lines: | 2830 |
| Uncovered lines: | 3973 |
| Coverable lines: | 6803 |
| Total lines: | 11960 |
| Covered branches: | 0 |
| Total branches: | 0 |
| **Method coverage:** | [Feature is only available for sponsors](https://reportgenerator.io/pro) |
| Tag: | 681_11970477973 |

</details>

## Coverage
<details><summary> - 35.3%</summary>

|**Name**|**Line**|**Branch**|
|:---|---:|---:|
|****|**35.3%**|****|
|b_stationarity_oracle|0%||
|build_unlifted_tf_alpha_expressions|62.5%||
|find_nonscalar|88.2%||
|generate_mpcc_relaxation_bounds|100%||
|generate_tf_inelastic_alphas|76.9%||
|get_psi_fun|86.8%||
|print_problem_details|0%||

</details>
<details><summary>nosnoc - 79%</summary>

|**Name**|**Line**|**Branch**|
|:---|---:|---:|
|**nosnoc**|**79%**|****|
|nosnoc.error|100%||
|nosnoc.Integrator|68.1%||
|nosnoc.Options|73.2%||
|nosnoc.warning|100%||

</details>
<details><summary>nosnoc.dcs - 59.9%</summary>

|**Name**|**Line**|**Branch**|
|:---|---:|---:|
|**nosnoc.dcs**|**59.9%**|****|
|nosnoc.dcs.Base|0%||
|nosnoc.dcs.Cls|71.4%||
|nosnoc.dcs.Gcs|69.3%||
|nosnoc.dcs.Heaviside|79%||
|nosnoc.dcs.Objects|0%||
|nosnoc.dcs.Stewart|80.3%||

</details>
<details><summary>nosnoc.discrete_time_problem - 38.6%</summary>

|**Name**|**Line**|**Branch**|
|:---|---:|---:|
|**nosnoc.discrete_time_problem**|**38.6%**|****|
|nosnoc.discrete_time_problem.Cls|32.9%||
|nosnoc.discrete_time_problem.Gcs|56.6%||
|nosnoc.discrete_time_problem.Heaviside|53.4%||
|nosnoc.discrete_time_problem.Objects|0%||
|nosnoc.discrete_time_problem.Stewart|61.4%||

</details>
<details><summary>nosnoc.integrator - 59.2%</summary>

|**Name**|**Line**|**Branch**|
|:---|---:|---:|
|**nosnoc.integrator**|**59.2%**|****|
|nosnoc.integrator.FESD|53.6%||
|nosnoc.integrator.Options|50%||
|nosnoc.integrator.SmoothedPss|73.4%||

</details>
<details><summary>nosnoc.model - 31.2%</summary>

|**Name**|**Line**|**Branch**|
|:---|---:|---:|
|**nosnoc.model**|**31.2%**|****|
|nosnoc.model.Base|42%||
|nosnoc.model.Cls|75.3%||
|nosnoc.model.Heaviside|7.1%||
|nosnoc.model.Objects|0%||
|nosnoc.model.Pds|76.9%||
|nosnoc.model.Pss|82.1%||

</details>
<details><summary>nosnoc.mpc - 0%</summary>

|**Name**|**Line**|**Branch**|
|:---|---:|---:|
|**nosnoc.mpc**|**0%**|****|
|nosnoc.mpc.Base|||
|nosnoc.mpc.FullMpcc|0%||
|nosnoc.mpc.Options|||

</details>
<details><summary>nosnoc.objects - 0%</summary>

|**Name**|**Line**|**Branch**|
|:---|---:|---:|
|**nosnoc.objects**|**0%**|****|
|nosnoc.objects.Ball|0%||
|nosnoc.objects.Counter|||
|nosnoc.objects.Ellipse|0%||
|nosnoc.objects.Object|0%||

</details>
<details><summary>nosnoc.ocp - 20.2%</summary>

|**Name**|**Line**|**Branch**|
|:---|---:|---:|
|**nosnoc.ocp**|**20.2%**|****|
|nosnoc.ocp.Solver|20.2%||

</details>
<details><summary>nosnoc.solver - 41.4%</summary>

|**Name**|**Line**|**Branch**|
|:---|---:|---:|
|**nosnoc.solver**|**41.4%**|****|
|nosnoc.solver.MpccMethod|||
|nosnoc.solver.mpccsol|100%||
|nosnoc.solver.MpccSolver|39.9%||
|nosnoc.solver.Options|68.7%||

</details>
<details><summary>nosnoc.solver.plugins - 19.2%</summary>

|**Name**|**Line**|**Branch**|
|:---|---:|---:|
|**nosnoc.solver.plugins**|**19.2%**|****|
|nosnoc.solver.plugins.Ipopt|64.1%||
|nosnoc.solver.plugins.Snopt|0%||
|nosnoc.solver.plugins.Uno|0%||
|nosnoc.solver.plugins.Worhp|0%||

</details>
<details><summary>nosnoc.time_freezing - 77.7%</summary>

|**Name**|**Line**|**Branch**|
|:---|---:|---:|
|**nosnoc.time_freezing**|**77.7%**|****|
|nosnoc.time_freezing.cls_elastic|97.7%||
|nosnoc.time_freezing.cls_inelastic|89.2%||
|nosnoc.time_freezing.cls_inelastic_multicontact|69.9%||
|nosnoc.time_freezing.reformulation|0%||

</details>

File: +nosnoc/+model/Objects.m (hunk 4 above)
Comment (apozharski) [reply +nosnoc/+model/Objects.m]: Removed. Padded Polytope related.

File: +nosnoc/+model/Objects.m (hunk 5 above)
Comment (apozharski) [reply +nosnoc/+model/Objects.m]: I am ok with rename but I don't really know what the preference is. GcsSdfObjects for the classes maybe?

File: +nosnoc/+objects/Ellipse.m [hunk 14]
```diff
@@ -0,0 +1,100 @@
+classdef Ellipse < nosnoc.objects.Object
```
Comment (nurkanovic) [inline +nosnoc/+objects/Ellipse.m:1]: maybe this name is too generic. you may want to highlight that it is related to a PDS  and name it PDSObject (not to confuse with PDSObjectS, ofc.)

File: examples/object_model/test_sim.m [hunk 15]
```diff
@@ -0,0 +1,51 @@
+clear all
```
Comment (nurkanovic) [inline examples/object_model/test_sim.m:1]: this script has not plot, or output and just tests an integrator.
there are several options:
1) remove it
2) make a real test
3) nicely plot output to make it more tangible

File: examples/pds_discs_examples/pass_discs_via_pdsobjects.m [hunk 16]
```diff
@@ -0,0 +1,95 @@
+% Optimal control example from
```
Comment (nurkanovic) [inline examples/pds_discs_examples/pass_discs_via_pdsobjects.m:1]: For fun I translated the pss_discs example now with the cool objects class.
It seems to work without big problems, however the ordering of the states in model.x does not follow the order how I define the ball but the order how the contacts are define, i.e. i get [x1,x3,x2] instead of [x1,x2,x3] (which i find more intutive)
This made later the cost weights and plots inconsistent, but the right order of x will fix all of this

**Comment by nurkanovic** [comment]
Thanks for the elaborate comments and addressing what I wrote before.
After fixing the integrator rebase bugs, and the minor comments I made after your last edit, this branch can be merged and the PR closed. :)

File: +nosnoc/+objects/Ellipse.m (hunk 14 above)
Comment (apozharski) [reply +nosnoc/+objects/Ellipse.m:1]: Imo this naming is fine actually because this infrastructure can be reused for the CLS version of SDF objects.

File: examples/pds_discs_examples/pass_discs_via_pdsobjects.m (hunk 16 above)
Comment (apozharski) [reply examples/pds_discs_examples/pass_discs_via_pdsobjects.m:1]: Yep now this should be  in "timestamp" i.e. order of creation.

**Review by nurkanovic — APPROVED** [review]
